pkg/logql/log/diff.go

```diff
@@ -0,0 +1,58 @@
+package log
+
+import (
+	"github.com/sergi/go-diff/diffmatchpatch"
+)
+
+// DiffFilter filters lines that are not different to the previous line.
+type DiffFilter struct {
+	// dmp is used to perform the diffing.
+	dmp *diffmatchpatch.DiffMatchPatch
+	// diff stores the diffs between the previous and current line.
+	diff []diffmatchpatch.Diff
+	// pretty determines whether the output is colored and prettified.
+	pretty bool
+	// prev is the previous line.
+	prev []byte
+}
+
+// Filter implements Filterer.
+// The line is only kept if there is a difference between the current line and the
+// previous line.
+// The first line passed to the expression is necessarily always kept.
+func (df *DiffFilter) Filter(line []byte) (keep bool) {
```

Comment:
Wouldn't it be more efficient to just use `[]byte == []byte` to test if there's a diff? I think the only time you need to _compute_ the diff is when it exists, which can be more efficiently checked by equality testing.

Reply:
Good point, will make that change.